tribe-ext-pdf-tickets/index.php

```diff
@@ -0,0 +1,876 @@
+		 * need to include any fonts (like the dejavu... ones) in
+		 * vendor/mpdf/ttfontdata
+		 *
+		 * @link https://mpdf.github.io/reference/mpdf-variables/overview.html
+		 * @link https://github.com/mpdf/mpdf/pull/490
+		 */
+		$mpdf = new mPDF( 'c' );
+
+		$mpdf->WriteHTML( $html );
+
+		return $mpdf;
+	}
+
+	/**
+	 * Get the full, current URL.
+	 *
+	 * @link https://css-tricks.com/snippets/php/get-current-page-url/
+	 *
+	 * @return string
+	 */
+	private function get_current_page_url() {
+		$url = @( $_SERVER["HTTPS"] != 'on' ) ? 'http://' . $_SERVER["SERVER_NAME"] : 'https://' . $_SERVER["SERVER_NAME"];
+		$url .= ( $_SERVER["SERVER_PORT"] !== '80' ) ? ":" . $_SERVER["SERVER_PORT"] : "";
+		$url .= $_SERVER["REQUEST_URI"];
+
+		return $url;
+	}
+
+	/**
+	 * Get the Attendee ID from the attempted URL.
+	 *
+	 * Once we have the current page URL, see if it looks like one of the PDF
+	 * Ticket file names, and, if it does, try to parse the Attendee ID out of
+	 * it. Should only be running when hitting a 404.
+	 *
+	 * @see Tribe__Extension__PDF_Tickets::uploads_directory_url()
+	 *
+	 * @param $url
+	 *
+	 * @return int
+	 */
+	private function get_attendee_id_from_attempted_url( $url ) {
+		$url = strtolower( esc_url( $url ) );
+
+		$beginning_url_check = $this->uploads_directory_url() . 'et_';
+
+		// bail if...
+		if (
+			// no URL
+			empty( $url )
+			// not looking for a file in Uploads that begins with 'et_'
+			|| false === $this->string_starts_with( $url, $beginning_url_check )
+			// not looking for a file ending in '.pdf'
+			|| false === $this->string_ends_with( $_SERVER['REQUEST_URI'], '.pdf' )
+		) {
+			return 0;
+		}
+
+		// remove from the front
+		$file_name = str_replace( $beginning_url_check, '', $url );
+
+		// look for an integer with double underscores on both sides
+		preg_match( '/__(\d+)__/', $file_name, $matches );
+
+		// bail if not found
+		if ( empty( $matches ) ) {
+			return 0;
+		}
+
+		$guessed_attendee_id = absint( $matches[1] );
+
+		if ( 0 >= $guessed_attendee_id ) {
+			return 0;
+		}
+
+		$post_type = get_post_type( $guessed_attendee_id );
+
+		if (
+			empty( $post_type )
+			|| false === $this->string_starts_with( $post_type, 'tribe_' )
+		) {
+			return 0;
+		}
+
+		// hopefully we were right!
+		return $guessed_attendee_id;
+	}
+
+	/**
+	 * Create and upload an 404'd PDF Ticket, then redirect to it
+	 * now that it exists.
+	 *
+	 * If we attempted to load a PDF Ticket but it was not found (404), then
+	 * create the PDF Ticket, upload it to the server, and reload the attempted
+	 * URL, adding a query string on the end as a cache buster and so the
+	 * 307 Temporary Redirect code is technically valid.
+	 *
+	 * @see Tribe__Extension__PDF_Tickets::get_current_page_url()
+	 * @see Tribe__Extension__PDF_Tickets::get_attendee_id_from_attempted_url()
+	 * @see Tribe__Extension__PDF_Tickets::do_upload_pdf()
+	 */
+	public function if_pdf_url_404_upload_pdf_then_reload() {
```

Comment:
This is a shaky error prone way of doing things. Typically we would have either a URL rewrite setup or just a URL query var. When that is in place we know they are trying to download a ticket, and no guesswork is involved. Then we also don't need these hacks. I'd strongly recommend that we rewrite this, and thus can remove the `get_current_page_url()` which also had issues and in particular remove `get_attendee_id_from_attempted_url()` which does all of the guessing.

Reply:
done